vermouth/gmx/itp_read.py

```diff
@@ -244,289 +208,187 @@ def _block(self, line, lineno=0):
+            atoms=treated_atoms,
+            parameters=parameters,
+            meta=meta,)
 
-        context = self.current_block
-        interaction_name = self.section[-1]
-        delete = False
+        interaction_list = context.interactions.get(section, [])
+        interaction_list.append(interaction)
+        context.interactions[section] = interaction_list
 
-        tokens = collections.deque(_tokenize(line))
 
-        n_atoms = self.interactions_natoms.get(interaction_name)
+    def _parse_block_atom(self, tokens, context):
 
-        _base_parser(
-            tokens,
-            context,
-            context_type=context_type,
-            section=interaction_name,
-            natoms=n_atoms,
-            delete=delete,
-            current_meta=self.current_meta
-            )
-   
-        # we need to reorder the atoms and interactions, because for virtual_sitesn
-        # the function type is on section position followed by an undefined number
-        # of atom indices from which the VS is built
- 
-        if '--' not in line:
-           first_atom = context.interactions['virtual_sitesn'][-1].atoms[0]
-           other_atoms = context.interactions['virtual_sitesn'][-1].atoms[2:]
-           atoms = [first_atom] + other_atoms
-   
-           sec_atom = context.interactions['virtual_sitesn'][-1].atoms[1]
-           params = [sec_atom]+context.interactions['virtual_sitesn'][-1].parameters
-       
-           context.interactions['virtual_sitesn'][-1].atoms[:]=atoms
-           context.interactions['virtual_sitesn'][-1].parameters[:]=params
-
-
-def _some_atoms_left(tokens, atoms, natoms):
-    """
-    Return True if the token list expected to contain atoms.
-
-    If the number of atoms is known before hand, then the function compares the
-    number of already found atoms to the expected number. If the '--' token if
-    found, it is removed from the token list and there is no atom left.
-
-    Parameters
-    ----------
-    tokens: collections.deque[str]
-        Deque of token to inspect. The deque **can be modified** in place.
-    atoms: list
-        List of already found atoms.
-    natoms: int or None
-        The number of expected atoms if known, else None.
-
-    Returns
-    -------
-    bool
-    """
-    if not tokens:
-        return False
-    if tokens and tokens[0] == '--':
-        tokens.popleft()
-        return False
-    if natoms is not None and len(atoms) >= natoms:
-        return False
-    return True
+        # deque does not support slicing
+        first_six = (tokens.popleft() for _ in range(6))
+        index, atype, resid, resname, name, charge_group = first_six
+        # since the index becomes the node name and all graphs start with 0
+        # index it makes more sense to also directly start at 0
 
+        if int(index) < 1:
+            msg = ('In section {} is a negative atom reference, which is not allowed.')
+            raise IOError(msg.format(section.name))
```

Comment:
I think I would decouple the atomid (number specified in the itp file) from the node index (the key in `block.nodes`)